lib/rules/no-dupe-disjunctions.ts

```diff
@@ -175,6 +200,90 @@ function toNFA(
     }
 }
 
+/**
+ * Yields all nested alternatives in the given root alternative.
+ *
+ * This will yield actual alternative nodes as well as character class
+ * elements. The elements of a non-negated character class (e.g. `[abc]`) can
+ * be thought of as an alternative. That's why they are returned.
+ */
+function* iterateNestedAlternatives(
+    alternative: Alternative,
+): Iterable<NestedAlternative> {
+    for (const e of alternative.elements) {
+        if (e.type === "Group" || e.type === "CapturingGroup") {
+            for (const a of e.alternatives) {
+                if (e.alternatives.length > 1) {
+                    yield a
+                }
+
+                yield* iterateNestedAlternatives(a)
+            }
+        }
+
+        if (e.type === "CharacterClass" && !e.negate) {
+            const nested: NestedAlternative[] = []
+            for (const charElement of e.elements) {
+                if (charElement.type === "CharacterClassRange") {
+                    const min = charElement.min
+                    const max = charElement.max
+                    if (min.value === max.value) {
+                        nested.push(charElement)
+                    } else if (min.value + 1 === max.value) {
+                        nested.push(min, max)
+                    } else {
+                        nested.push(charElement, min, max)
+                    }
+                } else {
+                    nested.push(charElement)
+                }
+            }
+            if (nested.length >= 1) yield* nested
```

Comment:
I prefer to use yield directly rather than using an array. 

```suggestion
            for (const charElement of e.elements) {
                if (charElement.type === "CharacterClassRange") {
                    const min = charElement.min
                    const max = charElement.max
                    if (min.value === max.value) {
                        yield charElement
                    } else if (min.value + 1 === max.value) {
                        yield min
                        yield max
                    } else {
                        yield charElement
                        yield min
                        yield max
                    }
                } else {
                    yield charElement
                }
            }
```

Reply:
I made a mistake here. The `nested.length >= 1` should have been `nested.length > 1`. I'll fix that.

We can't yield elements directly because we want to ensure that we yield _at least 2_ elements per character class. The same check is also done for array alternatives. 

```ts
if (e.alternatives.length > 1) {
    yield a
}
```

The "at least 2" condition is an optimization. I.e. in `a(b)c`, we don't have to check the `b` alternative because we know that the partial NFA of `b` will be the complete NFA of `a(b)c`, so we don't have to check `b`. 

The array is necessary for character classes because ranges may yield 1-3 elements, so it's not so easy to know how many elements we'll yield in the end.

Reply:
The optimization has been fixed and documented in 5f985a1fc88dcf5fb67ff1fbe09dc2f8d529bafe.

Reply:
I got it!